cmd/ctlog/verifyfulcio/main.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2022 The Sigstore Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	fulcioclient "github.com/sigstore/fulcio/pkg/api"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/signals"
+)
+
+type fulcios []string
+
+func (f *fulcios) String() string {
+	return fmt.Sprint(*f)
+}
+
+func (f *fulcios) Set(value string) error {
+	*f = append(*f, value)
+	return nil
+}
+
+var fulcioList fulcios
+
+type CertResponse struct {
+	Certificates []string `json:"certificates"`
+}
+
+func main() {
+
+	flag.Var(&fulcioList, "fulcio", "List of fulcios which must be in the list")
+	var ctlogURL = flag.String("ctlog-url", "ctlog.ctlog-system.svc", "CTLog to check Fulcios against.")
+	var ctlogPrefix = flag.String("log-prefix", "sigstorescaffolding", "Prefix to append to the gtlogURL url. This is basically the name of the log.")
+	flag.Parse()
+	var strictMatch = flag.Bool("strict", true, "If set to true ctlog must only contain the Fulcios in the list, no more, no less")
+	ctx := signals.NewContext()
+	fulcioURLs := make([]*url.URL, 0, len(fulcioList))
+	for _, f := range fulcioList {
+		u, err := url.Parse(f)
+		if err != nil {
+			logging.FromContext(ctx).Fatalf("Invalid fulcioURL %s : %v", f, err)
+		}
+		fulcioURLs = append(fulcioURLs, u)
+	}
+	fmt.Printf("GOT: %s\n", fulcioList.String())
+
+	// First grab the certs that CTLog has.
+	ctlog := fmt.Sprintf("%s/%s/ct/v1/get-roots", *ctlogURL, *ctlogPrefix)
+	ctlogResponse, err := http.Get(ctlog)
+	if err != nil {
+		logging.FromContext(ctx).Fatalf("Failed to get trusted certs from ctlog: %v", err)
+	}
+	defer ctlogResponse.Body.Close()
+	body, err := io.ReadAll(ctlogResponse.Body)
+	if err != nil {
+		logging.FromContext(ctx).Fatalf("Failed to read body from ctlog: %v", err)
+	}
+	certs := CertResponse{}
+	if err := json.Unmarshal(body, &certs); err != nil {
+		logging.FromContext(ctx).Fatalf("Failed to unmarshal body from ctlog: %v", err)
+	}
+	for i, cert := range certs.Certificates {
+		logging.FromContext(ctx).Infof("Got back cert %d: %s", i, cert)
+	}
+
+	// Keep track of certs found. Same index as fulcioURLs
```

Comment:
Can we guarantee that the order remains the same? It might be safer to build a map from cert serial number (which should be unique) to cert.

Reply:
Throughout the run of the script, yes. But, let me double-check :)

Reply:
Yes, because we do bytewise comparison, so it's the same index.